### PR TITLE
Implement interface for higher-order optimizers

### PR DESCRIPTION
**Pull request by fritzo**
This implements an interface compatible with higher-order optimizers, those that use not only gradient information but also some number of Hessian-vector products. The motivating optimizer is `Newton2d`. Additionally we've implemented wrappers for Pyro optimizers and PyTorch optimizers, and added a container class to apply different optimizers to different parameters.

Docs: http://fritzo.org/temp/pyro/html/optimization.html#module-pyro.optim.multi

## Remaining work

- [x] support differentiation through the optimization process
- [ ] expose `cov` return value from `newton_step_2d()`

## Tested

- added convergence tests for all new classes. tests run in under 2 seconds.

Pair coded with @eb8680

### CODE REVIEW

File: pyro/optim/multi.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        self.trust_radii = trust_radii
+
+    def step(self, loss, params):
+        updated_values = {}
+        for name, value in params.items():
+            trust_radius = self.trust_radii.get(name)
+            updated_value, cov = newton_step_2d(loss, value, trust_radius)
+            updated_values[name] = updated_value.detach()
+        for name, value in params.items():
+            with torch.no_grad():
```
Comment (fritzo) [inline pyro/optim/multi.py]: We'll need to avoid the `no_grad` to support differentiating through the optimization process, but I believe that will require avoiding in-place tensor update and instead using `ParamStore.replace_param()`. I'm not sure the cleanest way to do this.

File: pyro/optim/multi.py [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+        to infinite trust radius.
+    """
+    def __init__(self, trust_radii={}):
+        self.trust_radii = trust_radii
+
+    def step(self, loss, params):
+        updated_values = {}
+        for name, value in params.items():
+            trust_radius = self.trust_radii.get(name)
+            updated_value, cov = newton_step_2d(loss, value, trust_radius)
```
Comment (fritzo) [inline pyro/optim/multi.py]: We'll need to return this, say through a metadata dict return value from `MultiOptimizer.step()`

File: pyro/optim/multi.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        :class:`MultiOptimizer` object to be used for the named parameters.
+        Together the ``names`` should partition up all desired parameters to
+        optimize.
+    """
+    def __init__(self, parts):
+        self.parts = list(parts)
+
+    def step(self, loss, params):
+        for names_part, optim in self.parts:
+            optim.step(loss, {name: params[name] for name in names_part})
```
Comment (eb8680) [inline pyro/optim/multi.py]: Does this behave correctly when each optimizer takes multiple derivatives of `loss`?  Does it require special care from the user or optimizer, e.g. using `retain_graph=True` in `grad` or not calling `backward`?

File: pyro/optim/multi.py (hunk 2 above)
Comment (eb8680) [reply pyro/optim/multi.py]: Is `cov` meant to be used for a Laplace approximation?

File: pyro/optim/multi.py (hunk 1 above)
Comment (eb8680) [reply pyro/optim/multi.py]: Seems like we should just make it easier to update the param store as discussed yesterday, perhaps by adding a `force` flag or `set_tensor` kwarg to `param(...)`.

File: tests/optim/test_multi.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+import pyro
+import pyro.distributions as dist
+import pyro.optim
+import pyro.poutine as poutine
+from pyro.optim.multi import MixedMultiOptimizer, Newton2d, PyroMultiOptimizer, TorchMultiOptimizer
+from tests.common import assert_equal
+
+FACTORIES = [
+    lambda: PyroMultiOptimizer(pyro.optim.Adam({'lr': 0.05})),
+    lambda: TorchMultiOptimizer(torch.optim.Adam, {'lr': 0.05}),
```
Comment (eb8680) [inline tests/optim/test_multi.py]: Should we test more Pyro and Torch optimizers?  Are there any we'd expect to fail (LBFGS?)?

File: pyro/optim/multi.py (hunk 3 above)
Comment (fritzo) [reply pyro/optim/multi.py]: Yes this should behave correctly since we've taken care to set `create_graph=True` when calling `grad()` in `PyroMultiOptimizer`. All derived classes of `MultiOptimizer` will set `create_graph=True` so they are composable.

File: pyro/optim/multi.py (hunk 2 above)
Comment (fritzo) [reply pyro/optim/multi.py]: Yes `cov` can be interpreted as a Laplace approximation. In one application we use `cov` to filter out bad parameter values.

File: pyro/optim/multi.py (hunk 1 above)
Comment (eb8680) [reply pyro/optim/multi.py]: This PR is fine as-is but we should do that next

File: pyro/optim/multi.py [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+
+    def step(self, loss, params):
+        names = []
+        values = []
+        for name, value in params.items():
+            names.append(name)
+            values.append(value)
+        grads = torch.autograd.grad(loss, values, create_graph=True)
+        for x, g in zip(values, grads):
+            x.grad = g
```
Comment (fehiepsi) [inline pyro/optim/multi.py]: I think that the usual behavior of a Pytorch `optimizer` accumulates gradients, not set `.grad` like this. How about use [torch.autograd.backward](https://pytorch.org/docs/master/autograd.html#torch.autograd.backward) here instead? Something like this:
```
grads = torch.autograd.grad(loss, values, create_graph=True)
torch.autograd.backward(values, grads)
```

File: pyro/optim/multi.py [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+class PyroMultiOptimizer(MultiOptimizer):
+    """
+    Facade to wrap :class:`~pyro.optim.optim.PyroOptim` objects
+    in a :class:`MultiOptimizer` interface.
+    """
+    def __init__(self, optim):
+        self.optim = optim
+
+    def step(self, loss, params):
+        names = []
```
Comment (fehiepsi) [inline pyro/optim/multi.py]: I guess we don't need to store this list in this method.

Comment (fehiepsi) [reply pyro/optim/multi.py]: So I propose we use `params` as a list (not a dict) of parameters. :)

File: pyro/optim/multi.py (hunk 5 above)
Comment (fritzo) [reply pyro/optim/multi.py]: I think we cannot use `.backward()` because backward operates on the entire compute graph reachable (upstream from) the `loss` tensor, whereas we want to restrict to computing gradients only wrt the given variables. This is important for interleaving multiple optimizers via `MixedMultiOptimizer`. Does that seem correct to you?

File: pyro/optim/multi.py (hunk 6 above)
Comment (fritzo) [reply pyro/optim/multi.py]: After discussion with @eb8680 we think it would be cleaner to start using dictionaries rather than lists in `PyroOptim` in Pyro 0.3. I also find it cleaner to use dicts rather than lists when combining optimizers in `MixedMultiOptimizer`. Does this sound reasonable?

File: pyro/optim/multi.py (hunk 5 above)
Comment (fehiepsi) [reply pyro/optim/multi.py]: I think that `torch.autograd.backward(values, grads)` just operates on the param values, not the entire graph. According to Adam in Pytorch's slack autograd channel (you can search about `torch.autograd.backward` there), this method is slightly faster.

File: tests/optim/test_multi.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+    lambda: Newton2d(trust_radii={'z': 0.2}),
+    lambda: MixedMultiOptimizer([(['y'], PyroMultiOptimizer(pyro.optim.Adam({'lr': 0.05}))),
+                                 (['x', 'z'], Newton2d())]),
+]
+
+
+@pytest.mark.parametrize('factory', FACTORIES)
+def test_optimizers(factory):
+    optim = factory()
```
Comment (jpchen) [inline tests/optim/test_multi.py]: nit: why are you calling lambdas instead of just passing in the object?

File: pyro/optim/multi.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+
+    def step(self, loss, params):
+        values = params.values()
+        grads = torch.autograd.grad(loss, values, create_graph=True)
+        for x, g in zip(values, grads):
+            x.grad = g
+        self.optim(values)
+
+
+class TorchMultiOptimizer(PyroMultiOptimizer):
```
Comment (jpchen) [inline pyro/optim/multi.py]: what is the purpose of torch multi optimizer since it is essentially a pyro multi optimizer (`optim` is wrapped in `PyroOptim` during initialization anyway)?  is there a use case for this?

Comment (fritzo) [reply pyro/optim/multi.py]: This is intended to help us explore the possibility of moving towards `grad`-based optimizers rather than `backward`-based `PyroOptim` #628 . @eb8680 and I want to use this assess whether we can get by with the `TorchMultiOptimizer` interface and possibly deprecate the `PyroOptim` interface, moving `PyroOptim`'s logic into `TorchMultiOptimizer`. We do not intend to address whether we should make this refactoring, only whether it is possible.

File: pyro/optim/multi.py (hunk 5 above)
Comment (fritzo) [reply pyro/optim/multi.py]: `torch.autograd.backward(values, grads)` "Computes the sum of gradients of given tensors w.r.t. graph leaves.", whereas we want to compute gradients only w.r.t. a subset of the leaves. Thus `backward` is not compatible with `MixedMultiOptimizer`. Adam has commented that `grad` is negligibly faster than `backward`; it performs the same tensor ops and has to walk the autograd graph twice, however this walking is cheap.

Comment (fehiepsi) [reply pyro/optim/multi.py]: @fritzo I think that in this case, these `values` are leaves, so this method just accumulates the `grads` into their `.grad` attributes (no real `backward` is processed). However, I agree that if your purpose is to not accumulate (`x.grad = g` instead of `x.grad = x.grad + g`), then we don't need to use that hacking statement.

Comment (fritzo) [reply pyro/optim/multi.py]: I believe the `values` are not leaves but rather roots. By default `values = [loss]` and `grads = [torch.tensor(1.)]`

Comment (fehiepsi) [reply pyro/optim/multi.py]: I mean `values` and `grads` are `values` and `grads` in this statement `grads = torch.autograd.grad(loss, values, create_graph=True)` and these `values` are params. The code I have mentioned is
```
grads = torch.autograd.grad(loss, values, create_graph=True)
torch.autograd.backward(values, grads)  # not torch.autograd.backward([loss], [torch.tensor(1.)])
```
Sorry if I still miss something.
